### ISSUE TITLE
How to package analyzers?

### ISSUE THREAD

**Issue by evildour** [dotnet/roslyn#26222]
Hello, I would like to build Roslyn analyzers to be distributed with a C# library I work on. There seem to be two ways to do that:
1. Package the analyzers within the NuGet package for the library, at analyzers\dotnet\cs\ for example.
2. Package the analyzers in their own NuGet package (at the same path) and include that package as a package reference in the library.

Most examples I’ve seen (Roslyn included) seem to use option 2, but I see a few disadvantages (some minor) with this approach over option 1:
- Updating the analyzers takes more steps: with a separate package, you’d need to re-publish the new analyzer package, then update the reference to it in the target library, and then re-publish that library's package. By packaging them together, you only need to re-publish the one package.
- For projects including the NuGet package, Visual Studio shows the analyzers twice in the Dependencies node: once under Analyzers and a 2nd time under NuGet. When they are packaged together the analyzers are only under the Analyzers node.
- Users wanting to opt out of the analyzers need to include an extra reference (although this doesn’t seem to work currently anyway due to https://github.com/dotnet/sdk/issues/1212):
```
<PackageReference Include="Microsoft.CodeAnalysis" Version="2.7.0" />
<PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="1.1.0" ExcludeAssets="analyzers" />
```
as opposed to if they were in the same package:
```
<PackageReference Include="Microsoft.CodeAnalysis" Version="2.7.0" ExcludeAssets="analyzers" />
```

The only downside I can see with packaging everything together is if the analyzers and main library include assets in the same areas of the package, it would prevent someone from being able to selectively exclude just the assets from the analyzers. But if that’s isn’t the case, is there any other disadvantage I’m not seeing? If I picked option 1, is there anything that would prevent or make difficult a switch to option 2 in the future (splitting up the packages)?

I looked to the docs for guidance [here](https://docs.microsoft.com/en-us/nuget/reference/analyzers-conventions), but they weren’t that clear. They say “Analyzers themselves are typically _packaged and distributed **as part of** the NuGet packages_ that implement the API or library in question”, which would seem like option 1 is being recommended, but then it says “For a good example, see the System.Runtime.Analyzers package…”, which is a standalone package of analyzers - option 2. I think the docs should have some better guidance here.

Thank you.

**Comment by jmarolf** [comment]
@evildour the advantage of shipping them separately is that the user can upgrade to new versions of the analyzer without the need to change the version of the library they are referencing.

Suppose there is a bug in version `1.1.0` of `Microsoft.CodeAnalysis.Analyzers`.  If they are separate packages I just need to rev the version for `Microsoft.CodeAnalysis.Analyzers` to the one that has the bugfix.
```XML
<PackageReference Include="Microsoft.CodeAnalysis" Version="2.7.0" /> <!-- library dependencies stay the same -->
<!-- <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="1.1.0" /> old analyzers with bugs :( -->
<PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="1.2.0" /> <!-- new analyzers with bugfixes :) -->
```

If analyzers were just included in the package `Microsoft.CodeAnalysis` then a consumer needs to update their library dependencies and all the potential issues that brings.
```XML
<!--<PackageReference Include="Microsoft.CodeAnalysis" Version="2.7.0" /> old analyzers with bugs :( -->
<PackageReference Include="Microsoft.CodeAnalysis" Version="2.8.0" /> <!-- new analyzers with bugfixes :) but my product is now incompatible with older version of Visual Studio-->
```

In short: analyzers should be loosely coupled to the libraries they provide assistance for.

**Comment by evildour** [comment]
Ah, that's what I was missing. I still feel like the docs can be improved a bit here, but I'll close this. Thanks for the quick response!
